ilastik/applets/dataSelection/dataSelectionGui.py

```diff
@@ -515,21 +505,60 @@ def _determineLaneRange(self, infos: List[DatasetInfo], startingLaneNum=None):
 
         return (startingLaneNum, endingLane)
 
-    def _createDatasetInfos(self, roleIndex: int, filePaths: List[Path], rois=None):
-        """
-        Create a list of DatasetInfos for the given filePaths and rois
-        rois may be None, in which case it is ignored.
-        """
-        if rois is None:
-            rois = [None] * len(filePaths)
-        assert len(rois) == len(filePaths)
+    def _join_dataset_path(self, file_path: Path, internal_path: str) -> str:
+        """Return a proper path to the dataset in a file."""
+        p = file_path.absolute()
+        with contextlib.suppress(ValueError):
+            p = p.relative_to(self.topLevelOperator.WorkingDirectory.value)
 
+        prefix = str(p.as_posix())
+        suffix = internal_path.strip("/")
+        return f"{prefix}/{suffix}" if suffix else prefix
+
+    def _dataset_infos(self, role_index: int, file_paths: List[Path]) -> List[DatasetInfo]:
+        """Return combined DatasetInfo objects for all files."""
         infos = []
-        for filePath, roi in zip(filePaths, rois):
-            info = self._createDatasetInfo(roleIndex, filePath, roi)
-            infos.append(info)
+
+        for file_path in file_paths:
+            internal_paths = self._internal_paths(role_index, file_path)
+            if internal_paths is None:
+                return []
+
+            if len(internal_paths) == 1 and internal_paths[0]:
+                self._add_default_inner_path(roleIndex=role_index, inner_path=internal_paths[0])
+
+            infos += [
+                RelativeFilesystemDatasetInfo.create_or_fallback_to_absolute(
+                    filePath=self._join_dataset_path(file_path, internal_path),
+                    project_file=self.project_file,
+                    allowLabels=(self.guiMode == GuiMode.Normal),
+                )
+                for internal_path in internal_paths
+            ]
+
         return infos
 
+    def _internal_paths(self, role_index: int, file_path: Path) -> Optional[List[str]]:
+        """Return internal dataset paths for a file, possibly selected by a user.
+
+        If the file does not have internal datasets, return a singleton list with empty string.
+        If the user does not want to continue, return None.
+        """
+        if not DatasetInfo.fileHasInternalPaths(file_path):
+            return [""]
+
+        internal_paths = DatasetInfo.getPossibleInternalPathsFor(file_path)
+        if not internal_paths:
+            raise RuntimeError(f"File {file_path} has no image datasets")
+        if len(internal_paths) == 1:
+            return internal_paths
+
+        auto_internal_paths = list(self._get_previously_used_inner_paths(role_index).intersection(internal_paths))
+        if len(auto_internal_paths) == 1:
+            return auto_internal_paths
```

Comment:
If a user selects an .h5 file with multiple datasets, then picks one, that dataset will be recorded as an `auto_internal_path`. Say that the user then changes his mind, and removes that file from the data selection. Then, he tries to add it again, but this time he wants to add multiple datasets at once; He will not be able to, because `auto_internal_paths` will have his recorded selection from the first time he added the file, and this function will return before reaching `SubvolumeSelectionDlg`. In fact, he will never be able to use the multi-dataset selection for that file again, until he deletes the `.ilastik_preferences` file.

Reply:
Good catch! Could you explain why do we need to remember previously used paths anyway?

Reply:
hmm this was a "convenience function". Let's say a user has multiple h5 files with the same internal layout (always `/raw` and `/somethingelse`). We try to be helpful and if the user has selected the `raw` one for the first image, automatically load `raw` for the next file that is added (if `raw` is present).

Reply:
Maybe it would be better to add an "Apply to All" checkbox? We could make it checked by default to preserve the original behavior.

Also, how exactly do these selections persist between different actions? Do we save them in a project file?

Reply:
there are good arguments for either the possibilities you mention. There has been an [issue about it](https://github.com/ilastik/ilastik/issues/2129) where the fix results in the current behavior:

inner paths are only "remembered" for the current session with the current project. No other persistence. I'm not quite sure what you mean by "actions".

Reply:
> I'm not quite sure what you mean by "actions".
I meant a user action such as drag-and-drop, or clicking on the "Add..." button.

OK, so how users usually add files in ilastik? Do they frequently add/delete many different files that have multiple internal datasets, or load everything once in the beginning.

What if a user adds one multi-dataset file first, and then wants to add another multi-dataset file, but with different datasets? For me it would be very confusing if the system just goes and silently adds some previously selected datasets.

Reply:
> What if a user adds one multi-dataset file first, and then wants to add another multi-dataset file, but with different datasets? For me it would be very confusing if the system just goes and silently adds some previously selected datasets.

usually all files will bear the same internal structure in experiments. So I'd say this is the more common case. At least when this behavior was not in place we got complaints - I haven't heard/read any complaints about this behavior not being there.

However, I like your suggestion best, to have a checkbox like "remember internal path(s) for all following h5 files" or sometihng like that.